File: contrib/python/decode_network.py
#!/usr/bin/env python3
"""
decode_network.py: Python demonstration of libbdd C-ABI bindings for network protocol analysis.
Demonstrates:
- RFC 791 IPv4 20-byte base header parsing via 32-bit word patterns (4U4U6U2U16U, 16U3U13U, 8U8U16U)
- RFC 768 UDP 8-byte datagram header parsing via single 64-bit unit (16U16U16U16U)
- RFC 793 TCP 20-byte base segment parsing with sub-byte control flag extraction (4U3U1B1B1B1B1B1B1B1B1B16U)
- Bidirectional roundtrip packing with bdd.pack()
- Multi-packet capture stream dissection
"""

import os
import sys
import struct
import socket

# Ensure python/ directory is in sys.path
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
BASE_DIR = os.path.dirname(os.path.dirname(SCRIPT_DIR))
DATA_DIR = os.path.join(os.path.dirname(SCRIPT_DIR), "data")
sys.path.insert(0, os.path.join(BASE_DIR, "python"))

from bdd import Bdd

def format_ip(ip_int: int) -> str:
    """Format 32-bit integer as dotted-decimal IPv4 address."""
    return socket.inet_ntoa(struct.pack(">I", ip_int))

def decode_ipv4_header(bdd: Bdd, data: bytes):
    """Decode 20-byte IPv4 base header using libbdd patterns on 32-bit words."""
    if len(data) < 20:
        raise ValueError("IPv4 header requires at least 20 bytes")

    w0, w1, w2, w3, w4 = struct.unpack(">IIIII", data[:20])

    version, ihl, dscp, ecn, total_length = bdd.unpack("4U4U6U2U16U", w0)
    pkt_id, flags, frag_offset = bdd.unpack("16U3U13U", w1)
    ttl, protocol, checksum = bdd.unpack("8U8U16U", w2)
    src_ip = w3
    dst_ip = w4

    proto_name = {1: "ICMP", 6: "TCP", 17: "UDP"}.get(protocol, f"Unknown ({protocol})")
    df_flag = bool(flags & 0x2)
    mf_flag = bool(flags & 0x1)

    return {
        "version": version,
        "ihl": ihl,
        "header_len": ihl * 4,
        "dscp": dscp,
        "ecn": ecn,
        "total_length": total_length,
        "id": pkt_id,
        "flags": flags,
        "df": df_flag,
        "mf": mf_flag,
        "frag_offset": frag_offset,
        "ttl": ttl,
        "protocol": protocol,
        "proto_name": proto_name,
        "checksum": checksum,
        "src_ip": format_ip(src_ip),
        "dst_ip": format_ip(dst_ip),
    }

def decode_udp_header(bdd: Bdd, data: bytes):
    """Decode 8-byte UDP datagram header using a single 64-bit unit."""
    if len(data) < 8:
        raise ValueError("UDP header requires at least 8 bytes")

    u64 = struct.unpack(">Q", data[:8])[0]
    pattern = "16U16U16U16U"
    fields = bdd.unpack(pattern, u64)
    src_port, dst_port, length, checksum = fields

    # Verify roundtrip packing
    repacked = bdd.pack(pattern, fields)
    assert repacked == u64, "UDP roundtrip packing mismatch!"

    return {
        "src_port": src_port,
        "dst_port": dst_port,
        "length": length,
        "checksum": checksum,
    }

def decode_tcp_header(bdd: Bdd, data: bytes):
    """Decode 20-byte TCP base segment header with sub-byte flag extraction."""
    if len(data) < 20:
        raise ValueError("TCP header requires at least 20 bytes")

    w0, w1, w2, w3, w4 = struct.unpack(">IIIII", data[:20])

    src_port, dst_port = bdd.unpack("16U16U", w0)
    seq_num = w1
    ack_num = w2

    # Word 3: data_offset(4), reserved(3), 9 flags (1 bit each), window(16)
    offset, res, ns, cwr, ece, urg, ack, psh, rst, syn, fin, win = bdd.unpack(
        "4U3U1B1B1B1B1B1B1B1B1B16U", w3
    )
    checksum, urg_ptr = bdd.unpack("16U16U", w4)

    active_flags = []
    if cwr: active_flags.append("CWR")
    if ece: active_flags.append("ECE")
    if urg: active_flags.append("URG")
    if ack: active_flags.append("ACK")
    if psh: active_flags.append("PSH")
    if rst: active_flags.append("RST")
    if syn: active_flags.append("SYN")
    if fin: active_flags.append("FIN")

    return {
        "src_port": src_port,
        "dst_port": dst_port,
        "seq_num": seq_num,
        "ack_num": ack_num,
        "data_offset": offset,
        "header_len": offset * 4,
        "active_flags": active_flags,
        "syn": bool(syn),
        "ack": bool(ack),
        "psh": bool(psh),
        "fin": bool(fin),
        "window_size": win,
        "checksum": checksum,
        "urg_ptr": urg_ptr,
    }

def demo_standalone(bdd: Bdd):
    print("=================================================================")
    print(" 1. Standalone Network Header Decoding via libbdd")
    print("=================================================================")

    # 1.1 IPv4 Header
    ipv4_path = os.path.join(DATA_DIR, "sample_ipv4.bin")
    with open(ipv4_path, "rb") as f:
        ip_info = decode_ipv4_header(bdd, f.read())
    print("  --- IPv4 Header ---")
    print(f"    Version / IHL:    IPv{ip_info['version']}, {ip_info['header_len']} bytes")
    print(f"    Total Length:     {ip_info['total_length']} bytes")
    print(f"    Identification:   0x{ip_info['id']:04X} ({ip_info['id']})")
    print(f"    Flags:            DF={'Yes' if ip_info['df'] else 'No'}, MF={'Yes' if ip_info['mf'] else 'No'}")
    print(f"    TTL / Protocol:   {ip_info['ttl']} hops, {ip_info['proto_name']}")
    print(f"    Checksum:         0x{ip_info['checksum']:04X}")
    print(f"    Routing:          {ip_info['src_ip']} -> {ip_info['dst_ip']}")

    # 1.2 UDP Header
    udp_path = os.path.join(DATA_DIR, "sample_udp.bin")
    with open(udp_path, "rb") as f:
        udp_info = decode_udp_header(bdd, f.read())
    print("\n  --- UDP Header ---")
    print(f"    Source Port:      {udp_info['src_port']}")
    print(f"    Destination Port: {udp_info['dst_port']}")
    print(f"    Length:           {udp_info['length']} bytes")
    print(f"    Checksum:         0x{udp_info['checksum']:04X}")

    # 1.3 TCP Header
    tcp_path = os.path.join(DATA_DIR, "sample_tcp.bin")
    with open(tcp_path, "rb") as f:
        tcp_info = decode_tcp_header(bdd, f.read())
    print("\n  --- TCP Header ---")
    print(f"    Source Port:      {tcp_info['src_port']}")
    print(f"    Destination Port: {tcp_info['dst_port']}")
    print(f"    Seq / Ack:        {tcp_info['seq_num']} / {tcp_info['ack_num']}")
    print(f"    Header Length:    {tcp_info['header_len']} bytes")
    print(f"    Active Flags:     [{', '.join(tcp_info['active_flags'])}]")
    print(f"    Window Size:      {tcp_info['window_size']}")
    print(f"    Checksum:         0x{tcp_info['checksum']:04X}")

def demo_packet_stream(bdd: Bdd):
    print("\n=================================================================")
    print(" 2. Multi-Packet Capture Stream Dissection")
    print("=================================================================")
    packets_path = os.path.join(DATA_DIR, "sample_packets.bin")

    with open(packets_path, "rb") as f:
        stream = f.read()

    offset = 0
    pkt_num = 1

    while offset < len(stream):
        ip_info = decode_ipv4_header(bdd, stream[offset:])
        ip_len = ip_info["total_length"]
        pkt_bytes = stream[offset : offset + ip_len]
        header_len = ip_info["header_len"]
        payload_offset = header_len

        print(f"\n  Packet #{pkt_num} [{ip_len} bytes total]:")
        print(f"    IP Layer:     {ip_info['src_ip']} -> {ip_info['dst_ip']} ({ip_info['proto_name']})")

        if ip_info["protocol"] == 17:  # UDP
            udp_info = decode_udp_header(bdd, pkt_bytes[payload_offset:])
            print(f"    UDP Layer:    Port {udp_info['src_port']} -> {udp_info['dst_port']} (Length: {udp_info['length']}B)")
            dns_bytes = pkt_bytes[payload_offset + 8 :]
            if len(dns_bytes) >= 4:
                dns_u32 = struct.unpack(">I", dns_bytes[:4])[0]
                tx_id, flags = bdd.unpack("16U16U", dns_u32)
                print(f"    DNS Payload:  Query ID=0x{tx_id:04X}, Flags=0x{flags:04X}")

        elif ip_info["protocol"] == 6:  # TCP
            tcp_info = decode_tcp_header(bdd, pkt_bytes[payload_offset:])
            flags_str = ", ".join(tcp_info["active_flags"])
            print(f"    TCP Layer:    Port {tcp_info['src_port']} -> {tcp_info['dst_port']} [Flags: {flags_str}] Seq={tcp_info['seq_num']}")
            tcp_hdr_len = tcp_info["header_len"]
            app_payload = pkt_bytes[payload_offset + tcp_hdr_len :]
            if app_payload:
                preview = repr(app_payload.decode("latin1"))
                print(f"    App Payload:  {preview}")

        offset += ip_len
        pkt_num += 1

def main():
    bdd = Bdd()
    demo_standalone(bdd)
    demo_packet_stream(bdd)
    print("\nAll Python network protocol examples executed successfully!")

if __name__ == "__main__":
    main()

File: contrib/c/decode_network.c
/**
 * decode_network.c - C demonstration of libbdd native C-ABI for network packet analysis.
 *
 * Demonstrates:
 * - bdd_unpack_u64(): Parsing RFC 791 IPv4 20-byte base header (32-bit word slicing)
 * - bdd_unpack_u64(): Parsing RFC 768 UDP 8-byte datagram header (single 64-bit unit)
 * - bdd_unpack_u64(): Parsing RFC 793 TCP 20-byte base header with discrete 1-bit flags
 * - bdd_pack_u64()  : Bit-exact roundtrip repacking
 * - Dissecting raw multi-packet captures (sample_packets.bin)
 */

#include "bdd.h"
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <arpa/inet.h>

static uint32_t read_u32_be(const uint8_t *p) {
    return ((uint32_t)p[0] << 24) |
           ((uint32_t)p[1] << 16) |
           ((uint32_t)p[2] << 8)  |
           ((uint32_t)p[3]);
}

static uint64_t read_u64_be(const uint8_t *p) {
    uint64_t hi = read_u32_be(p);
    uint64_t lo = read_u32_be(p + 4);
    return (hi << 32) | lo;
}

static void ip_to_str(uint32_t ip, char *buf, size_t maxlen) {
    uint32_t net_ip = htonl(ip);
    inet_ntop(AF_INET, &net_ip, buf, maxlen);
}

static void demo_ipv4(const char *filepath) {
    printf("=================================================================\n");
    printf(" 1. IPv4 Header Parsing via libbdd C-ABI\n");
    printf("=================================================================\n");
    FILE *f = fopen(filepath, "rb");
    if (!f) {
        fprintf(stderr, "Failed to open %s\n", filepath);
        return;
    }
    uint8_t hdr[20];
    if (fread(hdr, 1, 20, f) != 20) {
        fprintf(stderr, "Failed to read 20 bytes from %s\n", filepath);
        fclose(f);
        return;
    }
    fclose(f);

    uint32_t w0 = read_u32_be(hdr);
    uint32_t w1 = read_u32_be(hdr + 4);
    uint32_t w2 = read_u32_be(hdr + 8);
    uint32_t src_ip = read_u32_be(hdr + 12);
    uint32_t dst_ip = read_u32_be(hdr + 16);

    uint64_t f0[5], f1[3], f2[3];
    bdd_unpack_u64("4U4U6U2U16U", w0, f0, 5);
    bdd_unpack_u64("16U3U13U", w1, f1, 3);
    bdd_unpack_u64("8U8U16U", w2, f2, 3);

    char src_str[INET_ADDRSTRLEN], dst_str[INET_ADDRSTRLEN];
    ip_to_str(src_ip, src_str, sizeof(src_str));
    ip_to_str(dst_ip, dst_str, sizeof(dst_str));

    const char *proto_name = (f2[1] == 17) ? "UDP" : ((f2[1] == 6) ? "TCP" : "Other");

    printf("  Input File:        %s\n", filepath);
    printf("  Version / IHL:     IPv%lu, %lu bytes (IHL: %lu)\n",
           (unsigned long)f0[0], (unsigned long)(f0[1] * 4), (unsigned long)f0[1]);
    printf("  Total Length:      %lu bytes\n", (unsigned long)f0[4]);
    printf("  Packet ID:         0x%04lX (%lu)\n", (unsigned long)f1[0], (unsigned long)f1[0]);
    printf("  Flags / Frag Off:  0x%lX [DF: %s] / %lu\n",
           (unsigned long)f1[1], (f1[1] & 0x2) ? "Yes" : "No", (unsigned long)f1[2]);
    printf("  TTL / Protocol:    %lu hops, %s (%lu)\n",
           (unsigned long)f2[0], proto_name, (unsigned long)f2[1]);
    printf("  Header Checksum:   0x%04lX\n", (unsigned long)f2[2]);
    printf("  Source IP:         %s\n", src_str);
    printf("  Destination IP:    %s\n", dst_str);

    // Roundtrip verification
    uint64_t repacked_w0 = 0;
    bdd_pack_u64("4U4U6U2U16U", f0, 5, &repacked_w0);
    printf("  Roundtrip w0 pack: 0x%08lX (%s)\n",
           (unsigned long)repacked_w0,
           (repacked_w0 == w0) ? "BIT-EXACT MATCH" : "MISMATCH");
}

static void demo_udp(const char *filepath) {
    printf("\n=================================================================\n");
    printf(" 2. UDP Datagram Header Parsing via libbdd C-ABI\n");
    printf("=================================================================\n");
    FILE *f = fopen(filepath, "rb");
    if (!f) {
        fprintf(stderr, "Failed to open %s\n", filepath);
        return;
    }
    uint8_t hdr[8];
    if (fread(hdr, 1, 8, f) != 8) {
        fclose(f);
        return;
    }
    fclose(f);

    uint64_t u64 = read_u64_be(hdr);
    uint64_t fields[4];
    bdd_unpack_u64("16U16U16U16U", u64, fields, 4);

    printf("  Input File:        %s\n", filepath);
    printf("  Source Port:       %lu\n", (unsigned long)fields[0]);
    printf("  Destination Port:  %lu%s\n",
           (unsigned long)fields[1], (fields[1] == 53) ? " (DNS)" : "");
    printf("  Datagram Length:   %lu bytes\n", (unsigned long)fields[2]);
    printf("  Checksum:          0x%04lX\n", (unsigned long)fields[3]);

    uint64_t repacked = 0;
    bdd_pack_u64("16U16U16U16U", fields, 4, &repacked);
    printf("  Roundtrip pack:    0x%016lX (%s)\n",
           (unsigned long)repacked,
           (repacked == u64) ? "BIT-EXACT MATCH" : "MISMATCH");
}

static void demo_tcp(const char *filepath) {
    printf("\n=================================================================\n");
    printf(" 3. TCP Base Segment Header with Discrete Sub-Byte Flags\n");
    printf("=================================================================\n");
    FILE *f = fopen(filepath, "rb");
    if (!f) {
        fprintf(stderr, "Failed to open %s\n", filepath);
        return;
    }
    uint8_t hdr[20];
    if (fread(hdr, 1, 20, f) != 20) {
        fclose(f);
        return;
    }
    fclose(f);

    uint32_t w0 = read_u32_be(hdr);
    uint32_t seq = read_u32_be(hdr + 4);
    uint32_t ack = read_u32_be(hdr + 8);
    uint32_t w3 = read_u32_be(hdr + 12);
    uint32_t w4 = read_u32_be(hdr + 16);

    uint64_t ports[2], flags_and_win[12], csum_urg[2];
    bdd_unpack_u64("16U16U", w0, ports, 2);
    bdd_unpack_u64("4U3U1B1B1B1B1B1B1B1B1B16U", w3, flags_and_win, 12);
    bdd_unpack_u64("16U16U", w4, csum_urg, 2);

    uint64_t offset = flags_and_win[0];
    uint64_t syn    = flags_and_win[9];
    uint64_t ack_f  = flags_and_win[6];
    uint64_t psh_f  = flags_and_win[7];
    uint64_t fin_f  = flags_and_win[10];
    uint64_t win    = flags_and_win[11];

    printf("  Input File:        %s\n", filepath);
    printf("  Source Port:       %lu\n", (unsigned long)ports[0]);
    printf("  Destination Port:  %lu%s\n",
           (unsigned long)ports[1], (ports[1] == 443) ? " (HTTPS)" : ((ports[1] == 80) ? " (HTTP)" : ""));
    printf("  Sequence Number:   %u\n", seq);
    printf("  Ack Number:        %u\n", ack);
    printf("  Header Length:     %lu bytes (Data Offset: %lu)\n",
           (unsigned long)(offset * 4), (unsigned long)offset);
    printf("  TCP Flags:         SYN=%lu ACK=%lu PSH=%lu FIN=%lu\n",
           (unsigned long)syn, (unsigned long)ack_f, (unsigned long)psh_f, (unsigned long)fin_f);
    printf("  Window Size:       %lu\n", (unsigned long)win);
    printf("  Checksum:          0x%04lX\n", (unsigned long)csum_urg[0]);
}

static void demo_stream(const char *filepath) {
    printf("\n=================================================================\n");
    printf(" 4. Multi-Packet Stream Capture Dissection\n");
    printf("=================================================================\n");
    FILE *f = fopen(filepath, "rb");
    if (!f) {
        fprintf(stderr, "Failed to open %s\n", filepath);
        return;
    }

    uint8_t pkt[256];
    int pkt_idx = 1;

    while (fread(pkt, 1, 20, f) == 20) {
        uint32_t w0 = read_u32_be(pkt);
        uint32_t w2 = read_u32_be(pkt + 8);
        uint32_t src_ip = read_u32_be(pkt + 12);
        uint32_t dst_ip = read_u32_be(pkt + 16);

        uint64_t f0[5], f2[3];
        bdd_unpack_u64("4U4U6U2U16U", w0, f0, 5);
        bdd_unpack_u64("8U8U16U", w2, f2, 3);

        uint64_t total_len = f0[4];
        uint64_t ihl_bytes = f0[1] * 4;
        uint64_t proto     = f2[1];

        char src_str[INET_ADDRSTRLEN], dst_str[INET_ADDRSTRLEN];
        ip_to_str(src_ip, src_str, sizeof(src_str));
        ip_to_str(dst_ip, dst_str, sizeof(dst_str));

        // Read remaining packet bytes
        size_t rem = (size_t)total_len - 20;
        if (fread(pkt + 20, 1, rem, f) != rem) break;

        const uint8_t *payload = pkt + ihl_bytes;

        printf("  Packet #%d [%lu bytes]: %s -> %s [Proto: %s]\n",
               pkt_idx++, (unsigned long)total_len, src_str, dst_str,
               (proto == 17) ? "UDP" : ((proto == 6) ? "TCP" : "Other"));

        if (proto == 17) {
            uint64_t udp_unit = read_u64_be(payload);
            uint64_t udp_fields[4];
            bdd_unpack_u64("16U16U16U16U", udp_unit, udp_fields, 4);
            printf("    UDP: Port %lu -> %lu (Len %luB)\n",
                   (unsigned long)udp_fields[0], (unsigned long)udp_fields[1], (unsigned long)udp_fields[2]);
        } else if (proto == 6) {
            uint32_t tcp_w0 = read_u32_be(payload);
            uint32_t tcp_seq = read_u32_be(payload + 4);
            uint32_t tcp_w3 = read_u32_be(payload + 12);
            uint64_t ports[2], f3[12];
            bdd_unpack_u64("16U16U", tcp_w0, ports, 2);
            bdd_unpack_u64("4U3U1B1B1B1B1B1B1B1B1B16U", tcp_w3, f3, 12);

            printf("    TCP: Port %lu -> %lu | Seq=%u | SYN=%lu ACK=%lu PSH=%lu\n",
                   (unsigned long)ports[0], (unsigned long)ports[1], tcp_seq,
                   (unsigned long)f3[9], (unsigned long)f3[6], (unsigned long)f3[7]);
        }
    }
    fclose(f);
}

static char *resolve_path(const char *rel_path) {
    char *buf = malloc(512);
    if (!buf) return NULL;

    FILE *f = fopen(rel_path, "rb");
    if (f) { fclose(f); strcpy(buf, rel_path); return buf; }

    snprintf(buf, 512, "contrib/%s", rel_path);
    f = fopen(buf, "rb");
    if (f) { fclose(f); return buf; }

    snprintf(buf, 512, "../%s", rel_path);
    f = fopen(buf, "rb");
    if (f) { fclose(f); return buf; }

    strcpy(buf, rel_path);
    return buf;
}

int main(int argc, char **argv) {
    char *ipv4_file    = (argc > 1) ? strdup(argv[1]) : resolve_path("data/sample_ipv4.bin");
    char *udp_file     = (argc > 2) ? strdup(argv[2]) : resolve_path("data/sample_udp.bin");
    char *tcp_file     = (argc > 3) ? strdup(argv[3]) : resolve_path("data/sample_tcp.bin");
    char *packets_file = (argc > 4) ? strdup(argv[4]) : resolve_path("data/sample_packets.bin");

    demo_ipv4(ipv4_file);
    demo_udp(udp_file);
    demo_tcp(tcp_file);
    demo_stream(packets_file);

    free(ipv4_file);
    free(udp_file);
    free(tcp_file);
    free(packets_file);

    printf("\nAll C network protocol examples completed successfully!\n");
    return 0;
}
